design/cassandra/design.md

```diff
@@ -0,0 +1,186 @@
+              - matchExpressions:
+                - key: failure-domain.beta.kubernetes.io/region
+                  operator: In
+                  values:
+                  - us-east-1
+                - key: failure-domain.beta.kubernetes.io/zone
+                  operator: In
+                  values:
+                  - us-east-1c
+      # Rook Common Type: ResourceSpec
+      # Resources declares the CPU,RAM resources of a single instance of the Rack.
+      resources:
+        - name: cassandra 
+          requests:
+            cpu: "2000m"
+            memory: "4Gi"
+          limits:
+            cpu: "2000m"
+            memory: "4Gi"
+        - name: sidecar
+          requests:
+            cpu: "200m"
+            memory: "200M"
+          limits:
+            cpu: "200m"
+            memory: "200M"
+  # Optional: sidecarImage overwrites the default image used
+  sidecarImage: 
+    repository: "rook.io/sidecar-cassandra"
+    tag: "0.1"
+```
+
+* The operator will create a StatefulSet for each (Datacenter, Rack) tupple.
+* In practice, Datacenter is usually mapped to a Region and Rack to an Availability Zone.
+
+### Major Pain Point: Stable Pod Identity
+
+A problem that every approach so far has, is dealing with loss of persistence.
+If a Cassandra instance loses its data (ie the underlying node fails, the disk dies).
+The replacing instance has to provide the IP of the instance it's replacing. Cassandra doesn't support stable hostnames as an identification method.
+However, IPs on Kubernetes are ephemeral and that leaves us with 2 options:
+
+* **Bookkeeping:** we do some kind of bookkeeping on Host-IDs (UUIDs uniquely identifying Cassandra instances) or IP addresses. This method has 2 major drawbacks:
+  1. Complexity
+  2. Subject to race-conditions in edge-case scenarios: since we don't know the instance's identity from the beginning, we have to query the instance to get it. If the instance loses its data (ie due to underlying node fail), then we have no way of knowing the identity of the instance.
+* **Sticky IPs Workaround / Service per Instance:** we create a `ClusterIP` Service for each Cassandra instance(Pod). This essentially provides us with a stable IP address for each instance. Since Cassandra uses IP addresses as the means of identification, this means that we don't need to do any bookkeeping. Also, we are immune from race-conditions, as we know our instance's identity beforehand. Some sceptical thoughts about this method:
+  1. Feels/Seems hacky
+  2. Uses too many services: most clusters have a /12 IP block to utilize, so we do not expect them to run out of IPs soon.
+  3. Performance Issues: since the default implementation of ClusterIPs is iptables this could be an issue with clusters in the 100s/1000s. However, an IPVS implementation of ClusterIP is [GA in 1.11](https://kubernetes.io/blog/2018/07/09/ipvs-based-in-cluster-load-balancing-deep-dive/) and works well even with 1000s of Services. 
+
```

Comment:
Currently I'm experiencing some issues running this setup (manually though) where the ClusterIP is not the same as the Node/PodID thus resulting in some confusion regarding the resolution of nodes both internally as well as to outside services. I think that this could potentially be worked around if we could bind the service address to the nodes, however, I don't think that is possible as of now. I'd be cautious with supporting this in the longer term and would need extensive testing to be trustworthy and that is beside the named caveats.

On the flip side, I still prefer this approach to the "manual" bookkeeping as I think that will bring a host of unforeseen issues.

Also just for reference sake, this would be a blessing for our use case:
https://github.com/kubernetes/kubernetes/issues/28969

Reply:
Thanks for your comments @Arqu !
I'll try answering some of your concerns:
1. If you provide the ClusterIP as the `broadcast_address` and the Pod IP as the `listen_address` it shouldn't exhibit any unpredictable behaviour. I have a manual setup like this up and working.
2. As for exposing things to the outside world, this will most likely be done though an Ingress. Generally, exposing Cassandra to services outside Kubernetes is an issue for every approach, so we should deal with this when the time comes. A simple workaround I can think of is to use a Cassandra driver that gets the data from the coordinator node and doesn't try to find the owner of the range.